### PR TITLE
Spike: research and create POC using NestJS

### PR DESCRIPTION
**Pull request by mgibson-casper**
### 🔥 Summary
Switching over from Express to use NestJS.

### 😤 Problem / Goals
-

### 🤓 Solution
-

### 🗒️ Additional Notes
-

### CODE REVIEW

File: nest-app/src/.eslintrc [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        "varsIgnorePattern": "^_",
+        "args": "after-used",
+        "argsIgnorePattern": "^_"
+      }
+    ],
+    "simple-import-sort/imports": "error",
+    "simple-import-sort/exports": "error",
+    "@typescript-eslint/ban-ts-comment": "off"
+  }
+}
```
Comment (hoffmannjan) [inline nest-app/src/.eslintrc:27]: maybe we can add a few more rules like `airbnb` ones?

File: nest-app/src/accounts/accounts.controller.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+import { AccountsService } from "./accounts.service";
+
+export class AccountByHashOrPublicKeyParamDtp {
+  @IsValidPublicKeyOrHash("hashOrPublicKey", {
+    message: "Not a valid public key or hash.",
+  })
+  @IsString()
+  public hashOrPublicKey: string;
+}
```
Comment (hoffmannjan) [inline nest-app/src/accounts/accounts.controller.ts:17]: sweet 🍬

File: nest-app/src/accounts/accounts.service.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    const stateRootHash = await jsonRpc.getStateRootHash();
+    const balance = await jsonRpc.getAccountBalance(stateRootHash, uref);
+
+    if (!balance) {
+      throw new ApiError(StatusCodes.NOT_FOUND, "Balance not found.");
+    }
+
+    return balance.toString();
+  }
+}
```
Comment (hoffmannjan) [inline nest-app/src/accounts/accounts.service.ts:38]: good that we don't cache this - I wouldn't do this even for security reasons

File: nest-app/src/app.service.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+@Injectable()
+export class AppService {
+  constructor(@Inject(CACHE_MANAGER) private readonly cacheManager: Cache) {}
+
+  async onModuleInit() {
+    await this.getStatus();
+  }
+
+  @Cron("*/30 * * * *")
```
Comment (hoffmannjan) [inline nest-app/src/app.service.ts:16]: awesome!

File: nest-app/src/blocks/blocks.controller.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  public sortBy: string;
+
+  @IsIn(["desc", "asc"])
+  @IsString()
+  @IsOptional()
+  public orderBy: Sort = "desc";
+}
+
+export class BlocksByHashOrHeightParamDtp {
+  @IsValidHash("hashOrHeight", { message: "Not a valid hash." })
```
Comment (hoffmannjan) [inline nest-app/src/blocks/blocks.controller.ts]: some of this error messages are the same in many places - maybe we can have an enum with all of the errors?

File: nest-app/src/peers/peers.service.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  @Cron("*/2 * * * *")
+  async handleCron() {
+    await this.fetchPeersRpc();
+  }
+
+  async fetchPeersRpc() {
+    // TODO: temporary - currently determining better fix
+    if (this.isFetchingPeers) {
+      return;
+    }
```
Comment (hoffmannjan) [inline nest-app/src/peers/peers.service.ts:26]: were there times when it took more then 2 minutes?

**Review by hoffmannjan — COMMENTED** [review]
It looks amazing, I just added a few comments worth discussing :)

File: nest-app/src/.eslintrc (hunk 1 above)
Comment (mgibson-casper) [reply nest-app/src/.eslintrc:27]: `airbnb` eslint rules? 😄

File: nest-app/src/accounts/accounts.service.ts (hunk 3 above)
Comment (mgibson-casper) [reply nest-app/src/accounts/accounts.service.ts:38]: Yup, totally 👍

File: nest-app/src/peers/peers.service.ts (hunk 6 above)
Comment (mgibson-casper) [reply nest-app/src/peers/peers.service.ts:26]: Nope, and I tested a bunch on the mainnet. I believe the longest it ever took was a few seconds. Although I do believe this is a decent solution that will allow us to block requests if it's taking too long.

File: nest-app/src/blocks/blocks.controller.ts (hunk 5 above)
Comment (mgibson-casper) [reply nest-app/src/blocks/blocks.controller.ts]: Oh good call, I've added an enum to the validators and used here and in all the other places needed.

File: nest-app/src/.eslintrc (hunk 1 above)
Comment (hoffmannjan) [reply nest-app/src/.eslintrc:27]: yeah exactly !

**Review by hoffmannjan — APPROVED** [review]

**Pull request merged**